<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Somme-des-1/N" data-toc-modified-id="Somme-des-1/N-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Somme des 1/N</a></span></li><li><span><a href="#Somme-des-1/N^i" data-toc-modified-id="Somme-des-1/N^i-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Somme des 1/N^i</a></span></li><li><span><a href="#Produit-des-sommes-de-1/N^n" data-toc-modified-id="Produit-des-sommes-de-1/N^n-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Produit des sommes de 1/N^n</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></div>

# Sommes infinies

A vérifier : $1+\frac{1}{2}+\frac{1}{3}+\frac{1}{4}+\frac{1}{5}+...=(1+\frac{1}{2}+\frac{1}{4}+\frac{1}{8}+...)(1+\frac{1}{3}+\frac{1}{9}+\frac{1}{27}+....)(1+\frac{1}{5}+\frac{1}{25}+\frac{1}{125}+...)...$


Sachant que la première somme est connue pour diverger (vers $+\infty$) 

Et que le produit en second (produit infini de sommes infinies de puissances inverses de nombres premiers seulement) diverge également vers $+\infty$ (est-ce une raison pour dire que les deux termes sont égaux ? non, mais bref ... \[c'est une autre histoire])

## Somme des 1/N

Commençons par calculer la somme des $\frac{1}{N}$, en s'arrétant quand l'ajout d'un $\frac{1}{N}$ ne change plus la somme.
 
*n.b. : pour que les calculs s'effectuent en un temps raisonnable, avec un peu de précision pour autant, on fera les calculs en 32 bits (seulement)*

In [55]:
import array

def versFloat(v):
    t = array.array('f')
    t.append(v)
    return t[0]

en 32 bits, par exemple, la différence entre 0.1 (codé sur 32 bits selon la norme IEEE 754) et la valeur 0.1 est de l'ordre de $10^{-9}$ (il y a 23+1 bits significatifs, 1 bit de signe, 8 bits d'exposant)

In [56]:
versFloat(0.1)-0.1

1.4901161138336505e-09

In [86]:
S=versFloat(1.0)
N=1
S_=versFloat(S+1.0/(N+1.0))
while S!=S_:
    S=S_
    S_=versFloat(S+1.0/(N+1))
    N=N+1
print("la somme vaut ",S,"au bout de ",N,"calculs (l'élément suivant est négligé par le calcul sur 32 bits)")

la somme vaut  15.903690338134766 au bout de  2097152 calculs (l'élément suivant est négligé par le calcul sur 32 bits)


La somme qui devait diverger vers $+\infty$ échoue à 15.9 (à peu près)

## Somme des 1/N^i

Calculons maintenant l'une des sommes $1+\frac{1}{N}+\frac{1}{N^2}+\frac{1}{N^3}+...$

In [87]:
def sommeDes1SurNPuisI(N):
    S=versFloat(1.0)
    M = N
    S_=versFloat(S+1.0/M)
    while S!=S_:
        S=S_
        M=M*N
        S_=versFloat(S+1.0/M)
    return S

Les premières sommes ne sont pas trop mauvaises :

In [88]:
sommeDes1SurNPuisI(2)

2.0

In [89]:
sommeDes1SurNPuisI(3)

1.5000001192092896

In [90]:
sommeDes1SurNPuisI(4)

1.3333332538604736

In [91]:
sommeDes1SurNPuisI(5)

1.2499998807907104

Mais au bout d'un certains temps, il est clair que la somme stagne à 1 .

In [92]:
sommeDes1SurNPuisI(17000000)

1.0

## Produit des sommes de 1/N^n

Calculons maintenant le produit infini des sommes précédentes.

Pour cela, il faut pouvoir lister les nombres premiers (on ne fait le produit que des sommes de 1/N^n avec N premier).

J'ajoute donc un petit test de primalité :

In [93]:
def estPremier(N):
  if N<=10:
    return (N==2)or(N==3)or(N==5)or(N==7)        
  if N%2 == 0 : 
    return False
  if N%3 == 0 : 
    return False
  if N%5 == 0 : 
    return False
  if N%7 == 0 : 
    return False
  VN = int(N**0.5)
  D = 11
  while D <= VN:
    if N%D == 0 : 
        return False
    if N%(D+2) == 0: 
        return False
    D = D+6
  return True   

Test rapide de la primalité :

In [72]:
estPremier(113)

True

In [73]:
estPremier(333)

False

In [94]:
S=sommeDes1SurNPuisI(2)
N=3
S_=S*sommeDes1SurNPuisI(N)
while S!=S_:
    S=S_
    N=N+1
    while not(estPremier(N)):
        N=N+1
    S_=S*sommeDes1SurNPuisI(N)
print("le produit vaut ",S,"au bout de ",N,"calculs (l'élément suivant est négligé par le calcul sur 32 bits)")

le produit vaut  30.118309435557457 au bout de  16777259 calculs (l'élément suivant est négligé par le calcul sur 32 bits)


## Conclusion

En 32 bits, avec ces calculs "simples" :
* $1+\frac{1}{2}+\frac{1}{3}+\frac{1}{4}+\frac{1}{5}+... = 16 \pm \epsilon$
* $(1+\frac{1}{2}+\frac{1}{4}+\frac{1}{8}+...)(1+\frac{1}{3}+\frac{1}{9}+\frac{1}{27}+....)(1+\frac{1}{5}+\frac{1}{25}+\frac{1}{125}+...)...=30 \pm \epsilon$

Ce n'est pas vraiment la même chose ... (mais on n'est pas si loin)

*remarque : on aurait pu faire de meilleurs calculs \[par exemple en regroupant les termes quand ils deviennent trop petits], est-ce que l'on serait arriver à montrer que ce sont des sommes infinies, peut-être, mais alors on aurait eu du mal à dire qu'elles sont vraiment égales.*